### DISCUSSION TITLE
Act warning on form submission with `user.type(textInput, '{Enter}')`

### DISCUSSION THREAD

**Discussion by axelboc** [testing-library/user-event#964] [Q&A]
I have a form with a text input, a submit button (`button[type="submit"]`) and an `onSubmit` event handler that triggers a bunch of state updates and effects.

If I write my test as follows, everything works fine:

```js
await user.type(textInput, 'foo');
await user.click(submitBtn);
expect(screen.getBySomething(...)).toBeVisible();
```

However, if I type `{Enter}` in the text field to submit the form instead of pressing the button, I get an act warning:

```js
await user.type(textInput, 'foo{Enter}');
expect(screen.getBySomething(...)).toBeVisible();
```

It seems that form submission on `Enter` is handled via the intermediate `keypress` event that `@testing-library/user-event` dispatches. Is it possible that `@testing-library/user-event` somehow forgets to wrap this form submission in `act`?

**Comment by ph-fritsche** [comment]
All events are dispatched per https://github.com/testing-library/user-event/blob/8215e2eebb13134cd0600cc8a2481725c16cd30d/src/event/dispatchEvent.ts#L6-L43

The `submit` event per `keypress` is triggered in
https://github.com/testing-library/user-event/blob/8215e2eebb13134cd0600cc8a2481725c16cd30d/src/event/behavior/keypress.ts#L25-L30

**Reply by axelboc** [reply]
I think I'm in the case where it dispatches a [click on the form's submit button](https://github.com/testing-library/user-event/blob/8215e2eebb13134cd0600cc8a2481725c16cd30d/src/event/behavior/keypress.ts#L24), but yeah.

**Reply by axelboc** [reply]
I should add that I configure `user-event` with `delay: null`.

**Reply by ph-fritsche** [reply]
Try to reproduce your problem in [Codesandbox](https://codesandbox.io/s/userevent-react-vbcvs?file=/src/App.test.js). Doing so often reveals the root of it just by isolating it.

> I should add that I configure `user-event` with `delay: null`.

Have you tried using the default?
Is it possible that your components asynchronously update state?

**Reply by axelboc** [reply]
I've managed to switch back to using the default `delay`. I'm now just calling `userEvent.setup()` and I'm still seeing the `act` warning when submitting via `{Enter}`. The test uses real timers. I'll try to make a repro if I get the time :+1:
